### PR TITLE
[BEAM-9198] BeamSQL aggregation analytics functionality

### PR DESCRIPTION
**Pull request by jhnmora000**
An concept-proof implementation for a cumulative sum using Analytic functions.

Implemented in: BeamAnalyticFunctionsExperimentTest.testOverCumulativeSum()

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Samza | Spark
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+          fieldValues.addAll(sortedRowsAsList.get(idx).getValues());
+          fieldValues.add(aggOutput);
+          Row ou = Row.withSchema(outputSchema).addValues(fieldValues).build();
+          out.output(ou);
+        }
+      }
+    };
+  }
+
+  private static DoFn<Row, Row> sortPartition(final FieldAggregation fieldAgg) {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:236]: I am wondering if this transform can be reused? https://github.com/apache/beam/blob/master/sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/SortValues.java

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:236]: It does not bad to start from this ParDo though. I posted that transform just for curiosity. Not required to use SortValues at this moment.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+      this.outputSchema = s;
+      this.aggFields = af;
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollectionList<Row> input) {
+      PCollection<Row> r = input.get(0);
+      for (FieldAggregation af : aggFields) {
+        org.apache.beam.sdk.schemas.transforms.Group.ByFields<Row> myg =
+            org.apache.beam.sdk.schemas.transforms.Group.byFieldIds(af.partitionKeys);
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:185]: This is nice to use schema's transforms.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.RelNode;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.Window;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.type.RelDataType;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexInputRef;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexLiteral;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Lists;
+
+public class BeamWindowRel extends Window implements BeamRelNode {
+  public BeamWindowRel(
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:52]: Interesting. I didn't know that there is no aggregate calls in WindowRel.

Do you know where those calls are defined?

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamAnalyticFunctionsExperimentTest.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                            2,
+                            "lettuce",
+                            "vegetable",
+                            10)
+                        .getRows()))
+            .setRowSchema(schema);
+    String sql =
+        "SELECT item, purchases, category, sum(purchases) over "
+            + "(PARTITION BY category ORDER BY purchases ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)"
+            + " as total_purchases  FROM PCOLLECTION";
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamAnalyticFunctionsExperimentTest.java:81]: Can you paste logical plan of this query in to comments? Just to help me better understand what Calcite produces at least on Logical plan level.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java (hunk 3 above)
Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:52]: O now I see. The call are in `Group`.

**Comment by jhnmora000** [comment]
Hi.
I forgot to push my latest changes.
These two tests are working now:
- BeamAnalyticFunctionsExperimentTest.testOverCumulativeSum()
- BeamAnalyticFunctionsExperimentTest.testSimpleOverFunction()

**Comment by amaliujia** [comment]
retest this please

**Comment by amaliujia** [comment]
@jhnmora000 

You can run `./gradlew check" to check style violations.

**Comment by jhnmora000** [comment]
retest this please

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamAnalyticFunctionsExperimentTest.java (hunk 4 above)
Comment (amaliujia) [reply sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamAnalyticFunctionsExperimentTest.java:87]: Thanks. Support PartitionBy and OrderBy is huge!

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.RelFieldCollation;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.RelNode;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.AggregateCall;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.Window;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.type.RelDataType;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rex.RexLiteral;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Lists;
+
+public class BeamWindowRel extends Window implements BeamRelNode {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:49]: Add java doc for classes (java doc means comments starts with `/*` and ends with `*/`

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+  @Override
+  public PTransform<PCollectionList<Row>, PCollection<Row>> buildPTransform() {
+    Schema outputSchema = CalciteUtils.toSchema(getRowType());
+    final List<FieldAggregation> analyticFields = Lists.newArrayList();
+    this.groups.stream()
+        .forEach(
+            anAnalyticGroup -> {
+              List<Integer> partitionKeysDef = anAnalyticGroup.keys.toList();
+              List<Integer> orderByKeys = Lists.newArrayList();
+              List<Boolean> orderByDirections = Lists.newArrayList();
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:69]: Add a test that includes DESC for ORDER BY?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+              List<Boolean> orderByNullDirections = Lists.newArrayList();
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:70]: Add a test or add a checkArgument to disable (either one works for me) for NULL last or NULL first in ORDER BY.

For NULL handling, depends on you, you can leave it for future PRs.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+                            fc.direction == RelFieldCollation.Direction.ASCENDING);
+                        orderByNullDirections.add(
+                            fc.nullDirection == RelFieldCollation.NullDirection.FIRST);
+                      });
+              int lowerB = Integer.MAX_VALUE; // Unbounded by default
+              int upperB = Integer.MAX_VALUE; // Unbounded by default
+              if (anAnalyticGroup.lowerBound.isCurrentRow()) {
+                lowerB = 0;
+              } else if (anAnalyticGroup.lowerBound.isPreceding()) {
+                // pending
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:85]: Better to use `checkArgument` to stop execution when you are seeing an unsupported case.

See: https://www.baeldung.com/guava-preconditions

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+              } else if (anAnalyticGroup.lowerBound.isFollowing()) {
+                // pending
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:87]: Same

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+              }
+              if (anAnalyticGroup.upperBound.isCurrentRow()) {
+                upperB = 0;
+              } else if (anAnalyticGroup.upperBound.isPreceding()) {
+                // pending
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:92]: Same

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+              } else if (anAnalyticGroup.upperBound.isFollowing()) {
+                // pending
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:94]: Same

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java (hunk 5 above)
Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:49]: Based on your implementation below, it will be really awesome that you add what is supported or what are constrains. Then you can gradually update this java doc once you add new features.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+                  .forEach(
+                      anAggCall -> {
+                        List<Integer> argList = anAggCall.getArgList();
+                        Schema.Field field =
+                            CalciteUtils.toField(anAggCall.getName(), anAggCall.getType());
+                        Combine.CombineFn combineFn =
+                            AggregationCombineFnAdapter.createCombineFn(
+                                anAggCall, field, anAggCall.getAggregation().getName());
+                        FieldAggregation fieldAggregation =
+                            new FieldAggregation(
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:109]: Even though you will stop executions on some unsupported cases, I think it is still ok to keep this class definition (e.g. no need to remove those unused field).

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+  }
+
+  @Override
+  public NodeStats estimateNodeStats(RelMetadataQuery mq) {
+    NodeStats inputStat = BeamSqlRelUtils.getNodeStats(this.input, mq);
+    return inputStat;
+  }
+
+  @Override
+  public BeamCostModel beamComputeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:171]: Add a function java doc to describe your choice of cost and why.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+    @Override
+    public PCollection<Row> expand(PCollectionList<Row> input) {
+      PCollection<Row> inputData = input.get(0);
+      Schema inputSchema = inputData.getSchema();
+      for (FieldAggregation af : aggFields) {
+        if (af.partitionKeys.isEmpty()) {
+          // This sections simulate a KV Row
+          // Similar to the output of Group.byFieldIds
+          // When no partitions are specified
+          Schema inputSch = inputData.getSchema();
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:197]: The mock key generation here could be simplified by, something like

```
              windowedStream
                  .apply(WithKeys.of("dummy"))
                  .apply(GroupByKey.create())
```

note that maybe do not for many schema manipulation.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+          inputData =
+              apply
+                  .apply(ParDo.of(uniquePartition(mockKeySchema, simulatedKeyValueSchema)))
+                  .setRowSchema(simulatedKeyValueSchema);
+        } else {
+          org.apache.beam.sdk.schemas.transforms.Group.ByFields<Row> myg =
+              org.apache.beam.sdk.schemas.transforms.Group.byFieldIds(af.partitionKeys);
+          inputData = inputData.apply("partitionBy", myg);
+        }
+        inputData =
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:216]: Can you skip sort transform when there is no order by? Also leave a comment to say migrate to SortValues transform in the future.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamCalcRule.java [hunk 16]
```diff
@@ -37,7 +40,20 @@ private BeamCalcRule() {
 
   @Override
   public boolean matches(RelOptRuleCall x) {
-    return true;
+    boolean hasRexOver = false;
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamCalcRule.java:43]: Add a comment to explain what this piece of code is doing. Basically it stops converting in the OVER clause case.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamWindowRule.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+import org.apache.beam.sdk.extensions.sql.impl.rel.BeamLogicalConvention;
+import org.apache.beam.sdk.extensions.sql.impl.rel.BeamWindowRel;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.plan.Convention;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.RelNode;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.convert.ConverterRule;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.Window;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.logical.LogicalWindow;
+
+public class BeamWindowRule extends ConverterRule {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamWindowRule.java:28]: Class java doc will be very helpful.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+          PCollection<Iterable<Row>> apply =
+              inputData.apply(org.apache.beam.sdk.schemas.transforms.Group.globally());
+          inputData =
+              apply
+                  .apply(ParDo.of(uniquePartition(mockKeySchema, simulatedKeyValueSchema)))
+                  .setRowSchema(simulatedKeyValueSchema);
+        } else {
+          org.apache.beam.sdk.schemas.transforms.Group.ByFields<Row> myg =
+              org.apache.beam.sdk.schemas.transforms.Group.byFieldIds(af.partitionKeys);
+          inputData = inputData.apply("partitionBy", myg);
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java:214]: In fact, if there is no order by but only partition by, combine per key will be the best API because backend can do combiner lift optimizations, which works as follow:

KV -> GroupByKey -> KV -> Combine
can be optimized as
KV -> local combine -> K CombinedV -> GroupByKey -> Combine all combined V.

So basically there could be a pre-combine before shuffle (GroupByKey), and after that each worker will only need to combine those pre-combined value (because of associativity rule).  This optimization will reduce lots of data through shuffle.

I think this can be left for a future work. Please log a JIRA to document this idea.

Comment (jhnmora000) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java]: I added a comment in the JIRA issue of the project [`issue`](https://issues.apache.org/jira/browse/BEAM-9198?focusedCommentId=17151728&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17151728)

**Comment by amaliujia** [comment]
retest this please.

File: sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamAnalyticFunctionsExperimentTest.java [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.Row;
+import org.junit.Test;
+
+/**
+ * A simple Analytic Functions experiment for BeamSQL created in order to understand the query
+ * processing workflow of BeamSQL and Calcite.
+ */
+public class BeamAnalyticFunctionsExperimentTest extends BeamSqlDslBase {
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/BeamAnalyticFunctionsExperimentTest.java:32]: Can you remove the "Experiment" from class. I am going to merge this PR so it won't be experiment anymore.

**Comment by amaliujia** [comment]
Before this PR is ready to be merged, please

1) run `./gradlew :sdks:java:extensions:sql:check` to check style violations and fix violations.
2) Can you rebase your PR against the head of master?

**Comment by jhnmora000** [comment]
retest this please

**Comment by amaliujia** [comment]
retest this please

**Comment by amaliujia** [comment]
Run Java PreCommit

**Comment by amaliujia** [comment]
Run Java PreCommit

**Comment by amaliujia** [comment]
Run Java PreCommit

**Comment by amaliujia** [comment]
@jhnmora000 thanks for you contribution!
